### PR TITLE
pre-commit: test PR74728

### PR DESCRIPTION
**Pull request by dtcxzyw**

### CODE REVIEW

File: bench/nlohmann_json/optimized/unit-regression2.cpp.ll [hunk 1]
```diff
@@ -87733,22 +87734,30 @@ cond.end.i.i:                                     ; preds = %if.end5
...
+cond.false14.i.i:                                 ; preds = %cond.end.i.i, %cond.end.thread.i.i
+  %sub647.i.i = phi i32 [ -150, %cond.end.thread.i.i ], [ %sub6.i.i, %cond.end.i.i ]
+  %add545.i.i = phi i64 [ %add537.i.i, %cond.end.thread.i.i ], [ %add5.i.i, %cond.end.i.i ]
+  %mul43.i.i = phi i64 [ %mul36.i.i, %cond.end.thread.i.i ], [ %mul.i.i, %cond.end.i.i ]
+  %v.sroa.6.042.i.i = phi i32 [ -149, %cond.end.thread.i.i ], [ %sub.i.i, %cond.end.i.i ]
+  %v.sroa.0.040.i.i = phi i64 [ %and.i.i, %cond.end.thread.i.i ], [ %add.i.i, %cond.end.i.i ]
+  %sub17.i.i = add nsw i64 %mul43.i.i, -1
+  br label %cond.end20.i.i
+
+cond.end20.i.i:                                   ; preds = %cond.false14.i.i, %cond.true8.i.i
```
Comment (dtcxzyw) [inline bench/nlohmann_json/optimized/unit-regression2.cpp.ll:87754]: Regression.

File: bench/openssl/optimized/libcrypto-lib-mem_sec.ll [hunk 2]
```diff
@@ -946,7 +946,9 @@ cond.false8:                                      ; preds = %sh_testbit.exit
   unreachable
 
 return:                                           ; preds = %sh_testbit.exit
-  ret i64 %shr.i14
+  %sh_prom = and i64 %list.0.lcssa.i, 2147483647
```
Comment (dtcxzyw) [inline bench/openssl/optimized/libcrypto-lib-mem_sec.ll:949]: Regression.

File: bench/openssl/optimized/libcrypto-lib-stack.ll [hunk 3]
```diff
@@ -479,22 +479,24 @@ if.then9:                                         ; preds = %if.end5
   br label %if.end22
 
 if.else:                                          ; preds = %if.end5
-  %3 = zext nneg i32 %loc to i64
-  %4 = getelementptr ptr, ptr %2, i64 %3
-  %arrayidx14 = getelementptr ptr, ptr %4, i64 1
+  %add = add nuw nsw i32 %loc, 1
```
Comment (dtcxzyw) [inline bench/openssl/optimized/libcrypto-lib-stack.ll:482]: Regression.

File: bench/qemu/optimized/ui_vnc-enc-tight.c.ll [hunk 4]
```diff
@@ -2186,7 +2186,8 @@ for.cond.preheader.i:                             ; preds = %for.end89.i, %if.el
   %x.080.i = phi i32 [ 0, %if.else62 ], [ %x.1.i, %for.end89.i ]
   %sub24.i = sub i32 %sub23.i, %x.080.i
   %sub.i31 = sub i32 %h, %y.081.i
-  %invariant.smin.i = tail call i32 @llvm.smin.i32(i32 %sub.i31, i32 %sub24.i)
+  %sub24.fr.i = freeze i32 %sub24.i
```
Comment (dtcxzyw) [inline bench/qemu/optimized/ui_vnc-enc-tight.c.ll:2189]: Regression.

File: bench/stb/optimized/stb_connected_components.c.ll [hunk 5]
```diff
@@ -657,11 +657,12 @@ for.body:                                         ; preds = %if.end39, %if.end12
   %indvars.iv75 = phi i64 [ %i.0, %if.end39 ], [ %indvars.iv.next76, %if.end124 ]
   %indvars.iv = phi i64 [ %j.0, %if.end39 ], [ %indvars.iv.next, %if.end124 ]
   %k.072 = phi i32 [ 0, %if.end39 ], [ %inc, %if.end124 ]
-  %7 = add nsw i64 %indvars.iv, %3
+  %7 = add nuw nsw i64 %indvars.iv, %3
+  %idxprom42 = and i64 %7, 4294967295
```
Comment (dtcxzyw) [inline bench/stb/optimized/stb_connected_components.c.ll:661]: Regression.

File: bench/stb/optimized/stb_herringbone_wang_tile.c.ll [hunk 6]
```diff
@@ -1389,8 +1389,9 @@ if.then12.us:                                     ; preds = %for.body6.us
   %7 = add nsw i64 %indvars.iv, %12
   %8 = mul nsw i64 %7, 3
   %arrayidx.us = getelementptr inbounds %struct.stbhw_tile, ptr %h, i64 0, i32 6, i64 %8
-  %9 = mul nsw i64 %6, 3
-  %add.ptr3.i.us = getelementptr inbounds i8, ptr %add.ptr.i.us, i64 %9
+  %9 = mul i64 %6, 3
+  %idx.ext2.i.us = and i64 %9, 4294967295
```
Comment (dtcxzyw) [inline bench/stb/optimized/stb_herringbone_wang_tile.c.ll:1393]: Regression.

File: bench/openssl/optimized/libcrypto-lib-stack.ll (hunk 3 above)
Comment (nikic) [reply bench/openssl/optimized/libcrypto-lib-stack.ll:482]: This is most likely fixed by https://github.com/llvm/llvm-project/commit/b8df88b41c8a1b4e879b4fd34be3522c9b45e86f.

File: bench/cvc5/optimized/inst_strategy_enumerative.cpp.ll [hunk 7]
```diff
@@ -886,6 +886,7 @@ ehcleanup192:                                     ; preds = %lpad172.split.us, %
   br label %eh.resume
 
 invoke.cont197:                                   ; preds = %if.end131
+  %inc211 = add nuw nsw i32 %r.0538, 1
```
[the remark this answers is not in the document]
Comment (nikic) [reply bench/cvc5/optimized/inst_strategy_enumerative.cpp.ll:889]: I'm not sure this one is really a regression. At least with the code in this form, I'm not sure we would consider sinking an add that is only needed on one path into the successor, even if it removes a phi node.

File: bench/stb/optimized/stb_connected_components.c.ll (hunk 5 above)
Comment (nikic) [reply bench/stb/optimized/stb_connected_components.c.ll:661]: I think to avoid this regression it might be necessary to represent the nneg flag in scev.

Comment (nikic) [reply bench/stb/optimized/stb_connected_components.c.ll:661]: Though it's very hard to understand what is actually going on from the end result -- my best guess is that we have something like zext (add i32) here originally and perform an sext IV widening and treat this as a truncating use, resulting in zext (trunc (add i64)). But I'm not confident this is the right guess.

Comment (nikic) [reply bench/stb/optimized/stb_connected_components.c.ll:661]: I think this demonstrates the issue:
```llvm
target datalayout = "n64"

define void @test(i32 %x, i32 %y, i32 %z) {
entry:
  br label %loop

loop:
  %iv = phi i32 [ 0, %entry ], [ %iv.inc, %loop ]
  %iv.sext = sext i32 %iv to i64
  call void @use(i64 %iv.sext)

  %add = add nuw nsw i32 %iv, %z
  %add.ext = zext nneg i32 %add to i64
  call void @use(i64 %add.ext)

  %iv.inc = add i32 %iv, 1
  %cmp = icmp ult i32 %iv, 100
  br i1 %cmp, label %loop, label %exit

exit:
 ret void
}

declare void @use(i64)
```

We perform an sext widening (apparently, we always prefer sext widening if there is at least one sext) and insert an unnecessary trunc after the add.

Comment (nikic) [reply bench/stb/optimized/stb_connected_components.c.ll:661]: This is most likely fixed by https://github.com/llvm/llvm-project/commit/c16559137cd1ad08ef5934558829d2ff5d227d70.

**Comment by nikic** [comment]
@dtcxzyw Can you please rerun the tests? I think the IndVars fix may also address most of the other regressions.

File: bench/icu/optimized/usearch.ll [hunk 8]
```diff
@@ -477,10 +477,12 @@ _ZL6getFCDPKDsPii.exit.i:                         ; preds = %if.end12.i.i.i, %lo
   br i1 %or.cond.i, label %land.lhs.true9.i, label %do.end.i
 
 land.lhs.true9.i:                                 ; preds = %_ZL6getFCDPKDsPii.exit.i
-  %arrayidx11.i = getelementptr i16, ptr %add.ptr2.i.i, i64 -2
-  %16 = load i16, ptr %arrayidx11.i, align 2
-  %17 = and i16 %16, -1024
-  %cmp14.i = icmp eq i16 %17, -10240
+  %16 = zext nneg i32 %2 to i64
```
Comment (dtcxzyw) [inline bench/icu/optimized/usearch.ll:480]: Regression.

**Pull request closed without merging**